cpp/modmesh/buffer/BufferExpander.hpp

```diff
@@ -70,19 +72,24 @@ class BufferExpander
     }
 
     BufferExpander(std::shared_ptr<ConcreteBuffer> const & buf, bool clone, ctor_passkey const &)
-        : m_concrete_buffer(clone ? buf->clone() : buf)
-        , m_begin(m_concrete_buffer->data())
-        , m_end(m_begin + m_concrete_buffer->size())
-        , m_end_cap(m_begin + m_concrete_buffer->size())
+        : BufferBase<BufferExpander>() // don't delegate m_begin and m_end, which will be overwritten later
+        , m_concrete_buffer(clone ? buf->clone() : buf)
     {
+        m_begin = m_concrete_buffer->data(); // override m_begin and m_end once we have the data
+        m_end = m_begin + m_concrete_buffer->size();
+        m_end_cap = m_begin + m_concrete_buffer->size();
     }
 
     BufferExpander(size_type nbyte, ctor_passkey const &)
+        : BufferBase<BufferExpander>() // don't delegate m_begin and m_end, which will be overwritten later
```

Comment:
This looks fishy.  IIUC, `expand()` needs `m_begin` to be valid in an execution path.  So `m_begin` needs to be initialized.

Reply:
no, `expand()` calls `reserve()`, which initialized `m_begin` and `m_end`.

Reply:
`reserve()` calls [`capacity()`](https://github.com/solvcon/modmesh/blob/17ce957c13acde8d40963be6b4b56af6bf0fa037/cpp/modmesh/buffer/BufferExpander.hpp#L103) which reads `m_begin` before initialization:

```cpp
    size_type capacity() const noexcept
    {
        return static_cast<size_type>(this->m_end_cap - this->m_begin);
    }
```

Reply:
thanks for catching this lol, I am surprised the test passed on macOS, seems it's a UB.